win,tty: fix read stop in line mode

orangemocha opened the pull request:
Closing the handle does not make `ReadConsoleW` exit reliably on Windows 7 and above. Thus, after switching from line to raw mode, keypresses were held until enter was pressed. This has caused a number of issues in node: https://github.com/nodejs/node/issues/2504 , https://github.com/nodejs/node/issues/2996, https://github.com/nodejs/node/issues/5384 , https://github.com/nodejs/node/issues/5821 , https://github.com/nodejs/node/issues/5940 .

This fix makes `ReadConsoleW` exit by writing a return keypress to its input buffer, similar to what was already done for raw mode. The second commit removes the duplicate handle that was used to close the console before. The third commit restores the cursor position to where it was before writing the enter, because `ReadConsoleW` echoes the newline in the screen.

This PR includes a test that switches the console to raw mode and uses `WriteConsoleInputW` to test reading in raw mode. I do not know of a unix equivalent to `WriteConsoleInputW`, so the test is for Windows only. In this test, switching to raw mode is done after a 100 ms sleep, to allow `uv_tty_line_read_thread` to reach [`ReadConsoleW`](https://github.com/libuv/libuv/blob/e397caa3a6a139fff2fed06e0223ee92582fa2d2/src/win/tty.c#L422).

The added `User32` library is necessary for [`MapVirtualKeyW`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms646306).

Fixes: libuv/libuv#852

cc @saghul @piscisaureus

eljefedelrodeodeljefe commented:
Who would have thought...

saghul commented on include/uv-win.h:
```diff
@@ -483,7 +483,6 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
   union {                                                                     \
     struct {                                                                  \
       /* Used for readable TTY handles */                                     \
-      HANDLE read_line_handle;                                                \
```
I think this will break the ABI because the size of `struct rd` is changed. Add a comment of it not being used anymore and we'll remove it for v2.

orangemocha replied:
Good call! Will fix.

saghul commented on src/win/tty.c:
```diff
@@ -57,11 +57,17 @@
 
 static void uv_tty_capture_initial_style(CONSOLE_SCREEN_BUFFER_INFO* info);
 static void uv_tty_update_virtual_window(CONSOLE_SCREEN_BUFFER_INFO* info);
+static int uv_cancel_read_console(uv_tty_t* handle);
```
Please make this `uv__cancel...`. I know the Windows side is not consistent with the rest, but let's try to make the new code consistent.

saghul commented on src/win/tty.c:
```diff
@@ -57,11 +57,17 @@
+static volatile BOOL uv_read_console_trap = FALSE;
+static volatile BOOL uv_read_console_started = FALSE;
+static HANDLE uv_read_console_done = NULL;
+
```
ditto.

saghul commented on src/win/tty.c:
```diff
@@ -463,25 +488,9 @@ static void uv_tty_queue_read_line(uv_loop_t* loop, uv_tty_t* handle) {
-
+  /* Reset flags */
+  uv_read_console_started = FALSE;
+  uv_read_console_trap = FALSE;
```
no locking here?

orangemocha replied:
I don't think it's necessary because no other thread can be accessing these at this point (we are starting a read and so there cannot be one in progress). It's also relying on the assumption that there is a memory barrier inside QueueUserWorkItem(). I'll add a comment. I could also add an explicit MemoryBarrier() call, or Enter/LeaveCriticalSection, but it will have a small performance penalty.

saghul commented on src/win/tty.c:
```diff
@@ -937,30 +949,109 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+    err = uv_cancel_read_console(handle);
+    if (err) {
+      return err;
+    }
```
drop braces for single statement ifs

saghul commented on src/win/tty.c:
```diff
@@ -937,30 +949,109 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
 
-  return 0;
+  /* Save screen state before sending the VK_RETURN event */
+  active_screen_buffer = CreateFileA("conout$",
```
is it possible that $conout doesn't point to the right thing? (redirected input or something)

orangemocha replied:
I don't think so. It's the documented way of getting active screen buffer handle per https://msdn.microsoft.com/en-us/library/windows/desktop/ms682088(v=vs.85).aspx.
In the worst case it will return INVALID_HANDLE_VALUE.

saghul commented on test/test-tty.c:
```diff
@@ -146,6 +146,75 @@ TEST_IMPL(tty) {
+  record.Event.KeyEvent.wVirtualScanCode = MapVirtualKeyW(VK_SPACE, MAPVK_VK_TO_VSC);
+  record.Event.KeyEvent.uChar.UnicodeChar = L' ';
+  record.Event.KeyEvent.dwControlKeyState = 0;
+  DWORD written;
```
declare this at the top.

saghul commented on uv.gyp:
```diff
@@ -112,6 +112,7 @@
               '-liphlpapi',
               '-lpsapi',
               '-lshell32',
+              '-luser32',
```
You also need to add this [here](https://github.com/libuv/libuv/blob/v1.x/configure.ac#L62) for MinGW.

saghul commented on src/win/tty.c:
```diff
@@ -57,11 +57,17 @@
+static CRITICAL_SECTION uv_read_console_trap_lock;
+static volatile BOOL uv_read_console_trap = FALSE;
+static volatile BOOL uv_read_console_started = FALSE;
+static HANDLE uv_read_console_done = NULL;
```
Maybe rename this to ... done_event?  Just "done" sounded boolean to me. Also, maybe initialize it to INVALID_HANDLE_VALUE?

saghul commented on src/win/tty.c:
```diff
@@ -109,6 +115,8 @@ static char uv_tty_default_inverse = 0;
 void uv_console_init() {
   InitializeCriticalSection(&uv_tty_output_lock);
+  InitializeCriticalSection(&uv_read_console_trap_lock);
+  uv_read_console_done = CreateEvent(NULL, 1, 0, NULL);
```
any chance this can fail? worth putting an assert?

bzoz replied:
Since  Vista it cannot fail: https://msdn.microsoft.com/library/windows/desktop/ms683472(v=vs.85).aspx

orangemocha replied:
InitializeCriticalSection won't fail but CreateEvent could. I'll add a check and an abort() in case of failure.

saghul commented on src/win/tty.c:
```diff
@@ -173,7 +181,6 @@ int uv_tty_init(uv_loop_t* loop, uv_tty_t* tty, uv_file fd, int readable) {
   if (readable) {
     /* Initialize TTY input specific fields. */
     tty->flags |= UV_HANDLE_TTY_READABLE | UV_HANDLE_READABLE;
-    tty->tty.rd.read_line_handle = NULL;
```
as mentioned above, initialize this to null and leave a comment that we need to remove it.

saghul commented:
Thanks a lot for working on this Alexis! I left some comments. There is one little important thing which definitely needs fixing: the ABI breakeage. Other than that, I don't have enough experience with low level Windows console handling to make a final judgement call, so I'll trust you and @piscisaureus.

orangemocha commented:
Thanks for the feedback, @saghul ! I have added 3 commits to address your comments. I can fixup the original commits once the PR has passed review.

saghul commented:
LGTM if the CI is green: https://ci.nodejs.org/job/libuv-test-commit-windows/48/ But I'd like @piscisaureus to take a look too, if possible.

TheAlphaNerd commented:
it looks like the libuv CI is down... this is likely related to the updated that came in yesterday. /cc @nodejs/build

saghul commented:
I just noticed, thanks for taking a look Myles!

saghul commented:
To save anyone a trip to the console:

~~~~
Started by upstream project "libuv-test-commit-windows" build number 50
originally caused by:
 Started by user Saúl Ibarra Corretgé
[EnvInject] - Loading node environment variables.
Building remotely on test-azure_msft-win2008r2-x64-2 (win2008r2-1p win2008r2 win2008r2-vs2015) in workspace c:\workspace\libuv-test-commit-windows\nodes\win2008r2
Wiping out workspace first.
Cloning the remote Git repository
Cloning repository https://github.com/$user/$project.git
 > git init c:\workspace\libuv-test-commit-windows\nodes\win2008r2 # timeout=10
Fetching upstream changes from https://github.com/$user/$project.git
 > git --version # timeout=10
 > git -c core.askpass=true fetch --tags --progress https://github.com/$user/$project.git +refs/heads/*:refs/remotes/origin/*
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git -c core.askpass=true fetch --tags --progress https://github.com/$user/$project.git +refs/heads/*:refs/remotes/origin/*" returned status code 128:
stdout: 
stderr: fatal: unable to access 'https://github.com/$user/$project.git/': The requested URL returned error: 400

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1719)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1463)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:63)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:314)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:506)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:120)
	at hudson.remoting.UserRequest.perform(UserRequest.java:48)
	at hudson.remoting.Request$2.run(Request.java:332)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at hudson.remoting.Engine$1$1.run(Engine.java:85)
	at java.lang.Thread.run(Unknown Source)
	at ......remote call to test-azure_msft-win2008r2-x64-2(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
	at hudson.remoting.Channel.call(Channel.java:781)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
	at sun.reflect.GeneratedMethodAccessor529.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
	at com.sun.proxy.$Proxy101.execute(Unknown Source)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1057)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1097)
	at hudson.scm.SCM.checkout(SCM.java:485)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1269)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:607)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
	at hudson.model.Run.execute(Run.java:1738)
	at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
ERROR: null
Notifying upstream projects of job completion
Finished: FAILURE
~~~~

TheAlphaNerd commented:
fingers crossed CI should be working now

ci: https://ci.nodejs.org/view/libuv/job/libuv-test-commit-windows/51/

I will keep this comment updated with the status of getting this working to avoid spamming the thread

saghul commented:
Thanks Myles! Build is 💚

orangemocha commented:
Great, thanks! I applied the fixups, should be ready for landing.

saghul commented:
Can you please reword the second commit? Now we are not removing anything :-)

I can work on a release tonight or on monday, but I guess nobody wants to release on a friday...

TheAlphaNerd commented:
@saghul we usually do the node releases on Tuesdays... so as long as it is out the door Monday I see no reason it could make it into next weeks v6 release

orangemocha commented:
Second commit reworded.

saghul commented:
Thanks Alexis!

@libuv/collaborators can I get another pair of eyes on this?
On May 13, 2016 11:25, "Alexis Campailla" <notifications@github.com> wrote:

> Second commit reworded.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/866#issuecomment-219007068>
>

piscisaureus commented on src/win/tty.c:
```diff
@@ -57,11 +57,17 @@
 static const uv_buf_t uv_null_buf_ = { 0, NULL };
 
+static CRITICAL_SECTION uv__read_console_trap_lock;
+static volatile BOOL uv__read_console_trap = FALSE;
```
There is no need to mark these volatile.

orangemocha replied:
You are probably right, but in my (long, possibly dated) experience with VC++ it's safer to rely on explicit volatile declarations. It should be noted that the volatile semantics are MS-specific: https://msdn.microsoft.com/en-us/library/12a04hfd.aspx.
It's possible that the compiler has gotten smarter and knows not to reorder accesses around critical sections, but is there really much harm in making it explicit?

piscisaureus commented on src/win/tty.c:
```diff
@@ -57,11 +57,17 @@
+static CRITICAL_SECTION uv__read_console_trap_lock;
+static volatile BOOL uv__read_console_trap = FALSE;
+static volatile BOOL uv__read_console_started = FALSE;
+static HANDLE uv__read_console_done_event = INVALID_HANDLE_VALUE;
```
I would suggest to make all of these part of struct uv_tty_y.

orangemocha replied:
On second thought, I am not sure I understand the suggestion about uv_tty_y, and I couldn’t find a similar pattern in libuv. Do you want me to put those global variables into a singleton struct instance?

piscisaureus commented on src/win/tty.c:
```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+
+  /* If ReadConsole already returned (the user pressed ENTER),
+     there is nothing to do */
+  if (WaitForSingleObject(uv__read_console_done_event, 0) == WAIT_OBJECT_0)
```
What if this thread gets to WaitForSingleObject before the 'reading' thread gets to ResetEvent ([there](https://github.com/janeasystems/libuv/blob/98548a05d5869ce31385efb3c70744ca1771a860/src/win/tty.c#L438))?

piscisaureus commented on src/win/tty.c:
```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+    err = 0;
   }
 
+  WaitForSingleObject(uv__read_console_done_event, INFINITE);
```
I would suggest not waiting for the reading thread to return. If you need a cue for restoring the cursor position, I think it'd be possible to make the reading thread do it right after it returns from ReadConsole?

orangemocha replied:
I agree it would be preferable not to wait. We were a bit concerned about what could happen with all the possible read_stop/read_start/set_mode sequences, and this way also minimizes the time between cursors jumps, but assuming nothing can modify the screen buffer state while UV_HANDLE_READ_PENDING is set, it should be doable. I'll take another look.

piscisaureus commented on src/win/tty.c:
```diff
@@ -463,25 +492,12 @@ static void uv_tty_queue_read_line(uv_loop_t* loop, uv_tty_t* handle) {
+     but we add an explicit one to be extra safe */
+  uv__read_console_started = FALSE;
+  uv__read_console_trap = FALSE;
+  MemoryBarrier();
```
Really unnecessary.

orangemocha replied:
Agreed. Everyone who reviewed the change before asked why there was no locking here, so I thought the explicit MemoryBarrier() would make it more readable. But the comment should be enough. I'll remove it.

piscisaureus commented on src/win/tty.c:
```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+    restore_screen_state = FALSE;
+  }
+
+  /* Write enter key event to force the console wait to return. */
```
I wonder if you considered writing anything else than Return. I believe ReadConsole() allows for specifying an alternative cue that the read operation needs to be interrupted, maybe there's something that has no visual side effects?

orangemocha replied:
Unfortunately, there isn't another way. I have brought this up to the team in Windows who works on the console APIs.

piscisaureus commented on src/win/tty.c:
```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+  if (!restore_screen_state)
+    goto out;
+
+  SetConsoleCursorPosition(active_screen_buffer, info.dwCursorPosition);
```
When the cursor is at the bottom of the console, it may cause the console buffer to scroll. In that situation restoring the cursor position would do more harm than good.

orangemocha replied:
Yes, this is a corner case that I know could be improved. At first I implemented this reusing the existing uv_tty_save_state/uv_tty_restore_state logic, but that had some apparent problems. After switching to the simpler approach, I haven't been able to produce any problems with scrolling in any of my tests, though I admit I was under time pressure to deliver this fix. Perhaps we can fix it in a second iteration? It should be a very rare case, and even a cursor jump is a minor issue compared to the current issue.

orangemocha commented:
@piscisaureus, I updated the last commit to address most of your feedback. PTAL. 
You can view the incremental fixup [here](https://github.com/janeasystems/libuv/commit/e966e04f38b7a6445dca609c140ad3843545aa66), should that make it easier for you to review.

I removed the wait from the main thread, and made the ReadConsole thread restore the console state. This should be safe and as you noted, it’s best not to have any waits in the main thread.

I also added some logic to handle the simple scroll case. If the cursor was at the bottom of the screen buffer, then we can assume that the ENTER we sent caused the screen buffer to scroll up of one line, so we adjust accordingly. Note that I am not handling the case in which the screen buffer scrolled up of more than one line since we called GetConsoleScreenBufferInfo, which would be hard to detect and is hopefully very unlikely in practice. Even if we hit that case, we would end up with the cursor one line below where it should be, which should essentially have the same effect of the ENTER being visible.

I also simplified the synchronization logic by using one status variable and interlocked operations.

saghul commented:
@piscisaureus I plan to release libuv with this fix (one you approve it) tonight, if you can get it reviewed by then it would be great.

[![](https://cdn.meme.am/instances/500x/63563997.jpg)]

piscisaureus commented on src/win/tty.c:
```diff
@@ -937,30 +976,80 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+  DWORD written;
+  DWORD err = 0;
+  LONG status;
+
```
assert(!(handle->flags & UV_HANDLE_CANCELLATION_PENDING))

orangemocha replied:
I can address this now

piscisaureus commented on src/win/tty.c:
```diff
@@ -937,30 +976,80 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
+  if (active_screen_buffer != INVALID_HANDLE_VALUE &&
+      GetConsoleScreenBufferInfo(active_screen_buffer,
+                                 &uv__saved_screen_state)) {
+    InterlockedOr(&uv__restore_screen_state, 1);
```
There's a bit of a race here in the sense that the ReadConsole operation could complete and check the restore_screen_state bit before it gets set here.

orangemocha replied:
We set this before sending the VK_ENTER, so ReadConsole won't have returned, unless the user pressed ENTER. If the user pressed ENTER, there can be a race between the user ENTER and the artificial ENTER, but at point either behavior should be ok.

piscisaureus commented:
> @piscisaureus, I updated the last commit to address most of your feedback. PTAL. 
You can view the incremental fixup here, should that make it easier for you to review.

> I removed the wait from the main thread, and made the ReadConsole thread restore the console state. This should be safe and as you noted, it’s best not to have any waits in the main thread.

Great!

> I also added some logic to handle the simple scroll case. If the cursor was at the bottom of the screen buffer, then we can assume that the ENTER we sent caused the screen buffer to scroll up of one line, so we adjust accordingly. Note that I am not handling the case in which the screen buffer scrolled up of more than one line since we called GetConsoleScreenBufferInfo, which would be hard to detect and is hopefully very unlikely in practice. 

> Even if we hit that case, we would end up with the cursor one line below where it should be, which should essentially have the same effect of the ENTER being visible.

Sounds good to me.

> I also simplified the synchronization logic by using one status variable and interlocked operations.

First of all, good job, it's much cleaner this way.
I think the amount of redundant state could be reduced even further, and the fact that the restore position needs another synchronized variable is a bit suspect to me.

But it's a the fine line between diligent and nitpicky, and many users are clamoring for a fix. 
So LGTM, go forth and land.

orangemocha commented:
[Added](https://github.com/libuv/libuv/pull/866/commits/5ad6688f5a3b5239be920fc9ed3eee9307359c42#diff-16746be892849b2c5daf1be3613ed28cR1014) the assert.

orangemocha commented:
One more CI run: https://ci.nodejs.org/job/libuv-test-commit-windows/57/

saghul commented:
CI is :green_heart: and Bert gave us his blessing, so I'm landing this.

saghul commented:
Landed in e51442bbc9b37dff32430c047844d9181798a9a8, 349aa6c0dd004960a44a83faf894e9d8592ca95b and 9eb131197171abda215452c5decd4a174ff1456d. Thanks Alexis and Joao for the patches, and Bert for the review!

orangemocha commented:
Awesome! Thanks a lot for the reviews and all the help, @saghul and @piscisaureus !
